src/rpc/service_eth.go

```diff
@@ -261,3 +274,51 @@ var DefaultProcessTransaction = func(tx *txtypes.Transaction, ctx context.Contex
 	}
 	return submitTx, nil
 }
+
+func (s *EthService) WaitTillMined(ctx context.Context, tx *types.Transaction, delayInSeconds int) error {
+	key, err := s.Cache.Key(tx)
+	if err != nil {
+		return err
+	}
+	value := s.Cache.InclusionCache[key]
+
+	if value == nil {
+		queryTicker := time.NewTicker(time.Duration(delayInSeconds) * 2 * time.Second)
+		defer queryTicker.Stop()
+		s.Cache.InclusionCache[key] = tx
+		utils.Logger.Info().Msgf("New tx recorded to check for inclusion | txHash: %s", tx.Hash().String())
+
+		for {
+			receipt, err := s.Processor.Client.TransactionReceipt(ctx, tx.Hash())
+			if err == nil {
+				delete(s.Cache.InclusionCache, key)
+				block, err := s.Processor.Client.BlockByHash(ctx, receipt.BlockHash)
+				if err != nil {
+					utils.Logger.Debug().Msgf("Error getting block | blockHash: %s", receipt.BlockHash.String())
+					return err
+				}
+				s.Processor.Db.FinaliseTx(db.TransactionDetails{
+					TxHash:        tx.Hash().String(),
+					InclusionTime: block.Time(),
+				})
+				return nil
+			}
+
+			if errors.Is(err, ethereum.NotFound) {
+				utils.Logger.Debug().Msgf("Transaction not yet mined | txHash: %s", tx.Hash().String())
+			} else {
+				delete(s.Cache.InclusionCache, key)
+				utils.Logger.Debug().Msgf("receipt retrieval failed | txHash: %s | err: %W", tx.Hash().String(), err)
+				return fmt.Errorf("receipt retrieval failed | txHash: %s | err: %W", tx.Hash().String(), err)
+			}
+
+			// Wait for the next round.
```

Comment:
we should stop waiting after a certain time not to wait indefinitely for transactions that are dropped, perhaps make this delay configurable